packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -440,44 +451,100 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
 
   void _handleMovement(
       LogicalKeyboardKey key, {
-      @required bool control,
+      @required bool wordModifier,
+      @required bool lineModifier,
       @required bool shift,
     }) {
+    if (wordModifier && lineModifier) {
+      // If both modifiers are down, nothing happens on any of the platforms.
+      return;
+    }
+
     TextSelection newSelection = selection;
 
     final bool rightArrow = key == LogicalKeyboardKey.arrowRight;
     final bool leftArrow = key == LogicalKeyboardKey.arrowLeft;
     final bool upArrow = key == LogicalKeyboardKey.arrowUp;
     final bool downArrow = key == LogicalKeyboardKey.arrowDown;
 
-    // Because the user can use multiple keys to change how they select, the
-    // new offset variable is threaded through these four functions and
-    // potentially changes after each one.
-    if (control) {
-      // If control is pressed, we will decide which way to look for a word
-      // based on which arrow is pressed.
-      if (leftArrow && newSelection.extentOffset > 2) {
-        final TextSelection textSelection = _selectWordAtOffset(TextPosition(offset: newSelection.extentOffset - 2));
-        newSelection = newSelection.copyWith(extentOffset: textSelection.baseOffset + 1);
-      } else if (rightArrow && newSelection.extentOffset < text.toPlainText().length - 2) {
-        final TextSelection textSelection = _selectWordAtOffset(TextPosition(offset: newSelection.extentOffset + 1));
-        newSelection = newSelection.copyWith(extentOffset: textSelection.extentOffset - 1);
+    // Returns true if [c] represents a whitespace code unit.
+    bool isWhitespace(int c) => (c <= 0x000D && c >= 0x0009) || c == 0x0020;
+
+    // Find the previous non-whitespace character
+    int previousNonWhitespace(int extent) {
+      int result = math.max(extent - 1, 0);
+      final String plain = text.toPlainText();
+      while (result > 0 && isWhitespace(plain.codeUnitAt(result))) {
+        result--;
       }
+      return result;
     }
-    // Set the new offset to be +/- 1 depending on which arrow is pressed
-    // If shift is down, we also want to update the previous cursor location
-    if (rightArrow && newSelection.extentOffset < text.toPlainText().length) {
-      newSelection = newSelection.copyWith(extentOffset: newSelection.extentOffset + 1);
-      if (shift) {
-        _cursorResetLocation += 1;
+
+    int nextNonWhitespace(int extent) {
+      final String plain = text.toPlainText();
+      int result = math.min(extent + 1, plain.length);
+      while (result < plain.length && isWhitespace(plain.codeUnitAt(result))) {
+        result++;
```

Comment:
nit: use += 1, https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#prefer-over